File: src/main/java/com/EmosewaPixel/pixellib/materialSystem/types/ObjectType.java
package com.EmosewaPixel.pixellib.materialSystem.types;

import com.EmosewaPixel.pixellib.materialSystem.lists.ObjTypes;
import com.EmosewaPixel.pixellib.materialSystem.materials.Material;
import net.minecraft.item.Item;
import net.minecraft.tags.ItemTags;
import net.minecraft.tags.Tag;
import net.minecraft.util.ResourceLocation;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;

/*
Object Types are objects used for generating blocks/items/fluids for certain materials.
This is the base class that shouldn't be used for generating anything
*/
public class ObjectType {
    private String name;
    private int volume = 0;
    private Predicate<Material> requirement;
    private List<String> tags = new ArrayList<>();

    public ObjectType(String name, Predicate<Material> requirement) {
        this.name = name;
        this.requirement = requirement;
    }

    public ObjectType addTypeTag(String... tags) {
        this.tags.addAll(Arrays.asList(tags));
        return this;
    }

    public ObjectType andRequires(Predicate<Material> requirement) {
        this.requirement = this.requirement.and(requirement);
        return this;
    }

    public ObjectType orRequires(Predicate<Material> requirement) {
        this.requirement = this.requirement.or(requirement);
        return this;
    }

    public ObjectType setBucketVolume(int mb) {
        volume = mb;
        return this;
    }

    public ObjectType build() {
        ObjTypes.add(this);
        return ObjTypes.get(name);
    }

    public String getName() {
        return name;
    }

    public int getBucketVolume() {
        return volume;
    }

    public boolean isMaterialCompatible(Material mat) {
        return requirement.test(mat);
    }

    public Tag<Item> getItemTag() {
        return new ItemTags.Wrapper(new ResourceLocation("forge", name + "s"));
    }

    public boolean hasTag(String tag) {
        return tags.contains(tag);
    }

    public List<String> getTypeTags() {
        return tags;
    }

    public void merge(ObjectType type) {
        if (type.getBucketVolume() != 0)
        volume = type.getBucketVolume();
        orRequires(type::isMaterialCompatible);
        tags.addAll(type.getTypeTags());
    }
}

File: src/main/java/com/EmosewaPixel/pixellib/materialSystem/types/ObjectType.kt
package com.EmosewaPixel.pixellib.materialsystem.types

import com.EmosewaPixel.pixellib.materialsystem.lists.ObjTypes
import com.EmosewaPixel.pixellib.materialsystem.materials.Material
import net.minecraft.item.Item
import net.minecraft.tags.ItemTags
import net.minecraft.tags.Tag
import net.minecraft.util.ResourceLocation

import java.util.ArrayList
import java.util.Arrays
import java.util.function.Predicate

/*
Object Types are objects used for generating blocks/items/fluids for certain materials.
This is the base class that shouldn't be used for generating anything
*/
open class ObjectType(val name: String, private var requirement: Predicate<Material>) {
    private var volume = 0
    private val tags = mutableListOf<String>()

    val itemTag: Tag<Item>
        get() = ItemTags.Wrapper(ResourceLocation("forge", name + "s"))

    val typeTags: List<String>
        get() = tags

    fun addTypeTag(vararg tags: String): ObjectType {
        this.tags.addAll(Arrays.asList(*tags))
        return this
    }

    fun andRequires(requirement: Predicate<Material>): ObjectType {
        this.requirement = this.requirement.and(requirement)
        return this
    }

    fun orRequires(requirement: Predicate<Material>): ObjectType {
        this.requirement = this.requirement.or(requirement)
        return this
    }

    fun setBucketVolume(mb: Int): ObjectType {
        volume = mb
        return this
    }

    fun build(): ObjectType {
        ObjTypes.add(this)
        return ObjTypes[name]!!
    }

    fun getBucketVolume() = volume

    fun isMaterialCompatible(mat: Material): Boolean {
        return requirement.test(mat)
    }

    fun hasTag(tag: String) = tag in tags

    fun merge(type: ObjectType) {
        if (type.getBucketVolume() != 0)
            volume = type.getBucketVolume()
        orRequires(Predicate { type.isMaterialCompatible(it) })
        tags.addAll(type.typeTags)
    }
}